File: Greedy/fractional-knapsack.py
# Algorithm: Fractional Knapsack Problem
# Type: Greedy Algorithm
# Time: O(n log n)
# Space: O(n)

def fractional_knapsack(values, weights, capacity):
    items = sorted(
        [(v / w, v, w) for v, w in zip(values, weights)],
        key=lambda x: x[0],
        reverse=True
    )

    total_value = 0
    for ratio, value, weight in items:
        if capacity >= weight:
            total_value += value
            capacity -= weight
        else:
            total_value += ratio * capacity
            break

    return total_value

File: Greedy/fractional-knapsack.c
// Algorithm: Fractional Knapsack Problem
// Type: Greedy Algorithm
// Time: O(n log n)
// Space: O(n)

#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int value, weight;
    float ratio;
} Item;

int compare(const void *a, const void *b) {
    return ((Item *)b)->ratio > ((Item *)a)->ratio ? 1 : -1;
}

float fractionalKnapsack(int W, Item items[], int n) {
    for (int i = 0; i < n; i++)
        items[i].ratio = (float)items[i].value / items[i].weight;

    qsort(items, n, sizeof(Item), compare);

    float totalValue = 0.0;
    for (int i = 0; i < n && W > 0; i++) {
        if (items[i].weight <= W) {
            totalValue += items[i].value;
            W -= items[i].weight;
        } else {
            totalValue += items[i].ratio * W;
            break;
        }
    }

    return totalValue;
}
